### ISSUE TITLE
Add namespace support for HashiCorp Vault

### ISSUE THREAD

**Issue by tmacam** [dapr/components-contrib#2806]
## Describe the feature

HashiCorp Vault component does not have support for [_namespaces_](https://developer.hashicorp.com/vault/docs/enterprise/namespaces), which are a feature used by SaaS solutions that offer cloud-hosted HashiCorp Vault services.

This was brought up in dapr/dapr#4779 and in Discord.

## Release Note

RELEASE NOTE: **ADD** Support for "namespaces" for HashiCorp Vault Component.

**Comment by Kreestyan** [comment]
Thank you for posting this as an issue @tmacam.
Indeed we can retrieve our secret from the SaaS using this curl request:
`curl --location "https://myvaultsaas.secretmgmt.mycompany.com/v1/MySecretPath/data/MySecret" --header "X-Vault-Token: s…f" --header "X-Vault-Namespace: MyNameSpace"`
But I don't see anywhere in the hashicorp secret store component, a way to reference this **_Vault-Namespace_**

To further document... 
Our Hashicorp provider created a temporary token given us access to their root name space and using this token I was able to retrieve their tests secrets using the dapr api:
`curl http://locahost:3500/v1.0/secrets/hashicorp/bulk`

However trying to use the dapr api to our assigned Vault-Namesapce = MyNameSpace results in this error:
`status code 403, body {\"errors\":[\"1 error occurred:\\n\\t* permission den
ied\\n\\n\"]}\n"}`
Exactly the same error as if I was omitting to provide header X-Vault-Namespace in my native curl request above

So basically the feature would just allow adding into the header "X-Vault-Namespace: <_vault namespace from component store_>"

Thank you!

**Comment by tmacam** [comment]
The main challenge that I foresee is the ability to test this feature. Vault is a stable component, so any new field must be tested in its certification tests. Unfortunately, the documentation for [namespaces](https://developer.hashicorp.com/vault/docs/enterprise/namespaces) make it seem like this is a cloud-only feature, so I am unsure we will be able to test this properly using the OSS Vault server.

I can think of ways to test this regardless (forcing HTTP and bringing up a python HTTP server just to verify this header) but this would be far from ideal.

**Comment by Kreestyan** [comment]
> The main challenge that I foresee to implement this feature is the ability to test it appropriately. Vault is a stable component, so any new field must be tested in its certification tests. Unfortunately, the documentation for [namespaces](https://developer.hashicorp.com/vault/docs/enterprise/namespaces) make it seem like this is a cloud-only feature, so I am unsure how we will be able to test this properly using the OSS Vault server.
> 
> I can think of ways to test this regardless (forcing HTTP and bringing up a python HTTP server just to verify this header) but this would be far from ideal.

I understand this might not be a bullet proof test but my organization can offer to test this feature as soon as it is in an alpha stage. Our provider could create additional test namespaces and we could test it against different namespaces, using different policies and/or secret path. Then on your end, you could take care of the regression testing to ensure no negative impact on existing features.

Any thing we can do to help, please let me know. We really need this feature otherwise, we will start implementing a native approach and I won't be able to justify rework just to use dapr.

**Comment by Taction** [comment]
@tmacam For the vault certificate test, it creates a hashicrop vault instance using docker, and then you can set the test namespaced value and get it using dapr. [EnginePath test is an example](https://github.com/dapr/components-contrib/blob/master/tests/certification/secretstores/hashicorp/vault/vault_test.go#L278).

For namespace support, from my point of view, it would be better to have component-level metadata as well that can be overwritten by request-level metadata. For each application, its namespace is more likely to be a fixed one.
For example, if I use the following component configuration, dapr adds `MyNameSpace` namespace for every request. And users can overwrite this value by sending a request with `namespace` metadata.
```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: vault
spec:
  type: secretstores.hashicorp.vault
  version: v1
  metadata:
  - name: namespace
    value: "MyNameSpace"
# ...
```
WDYT?

**Comment by Kreestyan** [comment]
I was hoping for the same. I'd rather not expose the vault namespace to the application as this would be specific to Hashicorp.

Also the existing API seems to already provide a metadata override for the namespace (see example [HERE](https://docs.dapr.io/reference/api/secrets_api/#examples))

But I tried it in our case, and it didn't work so I don't think that "namespace" is the Vault Namespace.

Therefore that new attribute should be very explicit and ensure we do not confuse them

**Comment by tmacam** [comment]
Thanks @tmacam  and @Taction 

> For namespace support, from my point of view, it would be better to have component-level metadata as well that can be overwritten by request-level metadata. (For each application, its namespace is more likely to be a fixed one.)

I agree with the suggestion of having it as a component-level metadata. I think that given how this is exposed and vended by cloud-providers, it makes perfect sense to expose this setting at a component-level.

As for having this configurable on a per-request basis, could you think of a use-case where this (per-request) would be required instead of the component-level route?

**Comment by Kreestyan** [comment]
Great to hear we are on the same page. As far as I am concerned, I do not see a use case where that parameters could be overridden at the application level because to me it's a vault host specific parameter and doing so would break the agnosticism  of the API.

The name of the attribute should ensure we differentiate it from the namespace of the local cluster, so for instance it could be vault-namespace or even vault-sas-namespace

I feel bad asking but any ETA for an alpha version? Then again, we can help for testing – as our implementation is blocked for now.



And also, not sure if this is the right place but how does this change would fit with the current option of providing a namespace in the metadata like in the [documented example](https://github.com/dapr/components-contrib/issues/curl%20http:/localhost:3500/v1.0/secrets/vault/db-secret?metadata.version_id=15&?metadata.namespace=production):
`curl http://localhost:3500/v1.0/secrets/vault/db-secret?metadata.version_id=15&?metadata.namespace=production`

_which BTW I think this example has a typo since it shows 2 ‘**?**’ where the 2nd one should be replaced by ‘**&**’_

**Comment by Taction** [comment]
> As for having this configurable on a per-request basis, could you think of a use-case where this (per-request) would be required instead of the component-level route?

@tmacam The main concern is that users may have multi namespaces for one application. But I am not an end user, I am inclined to have a component level that is enough. We needn't support per-request level metadata unless there is any end users ask to.

As @Kreestyan seems to be an end user, is there any use case that you will use different namespaces in the same application?

**Comment by Kreestyan** [comment]
> As @Kreestyan seems to be an end user, is there any use case that you will use different namespaces in the same application?
We only need it to be defined at the secret component store level. I do not see any use case where this value would have to be overridden by the application.

**Comment by tmacam** [comment]
> The name of the attribute should ensure we differentiate it from the namespace of the local cluster, so for instance it could be vault-namespace or even vault-sas-namespace

Good point. And prefixing with `vault` would be consistent with other vault properties naming.

> I feel bad asking but any ETA for an alpha version? Then again, we can help for testing – as our implementation is blocked for now.

I don't think this will make to 1.11. But if I made it in a branch of mine could you test it?

> And also, not sure if this is the right place but how does this change would fit with the current option of providing a namespace in the metadata like in the [documented example](https://github.com/dapr/components-contrib/issues/curl%20http:/localhost:3500/v1.0/secrets/vault/db-secret?metadata.version_id=15&?metadata.namespace=production): `curl http://localhost:3500/v1.0/secrets/vault/db-secret?metadata.version_id=15&?metadata.namespace=production`
 
> _which BTW I think this example has a typo since it shows 2 ‘**?**’ where the 2nd one should be replaced by ‘**&**’_

Could you cut an issue under dapr/docs for this and cc me to it? I think there are multiple issues with that document honestly:

* The multiple `?` in that URL. That is wrong. :P 
* That example URL is problematic in multiple ways. I _guess_ it is a made up example that was not even tested and I _guess_ its intent was to show _how_ to provide metadata using the HTTP interface to a hypotetical secret store that supported those features (namespace and versions). But it is an unfortunate one because it is showing features that no secret store currently support at the same time.
   * We do not support namespaces on vault. But we do support versioning on vault.
   * We _do_ [support namespace in k8s secrets](https://github.com/dapr/components-contrib/blob/22fea6f43d53a6dc4780132c8849577928e6ebcf/secretstores/kubernetes/kubernetes.go#L35), but we do not support _versions_** in k8s secrets.

**Comment by tmacam** [comment]
So it could be something as simple as 2808, which is available on my branch on https://github.com/tmacam/dapr-components-contrib/tree/VaultNamespace. Now: I haven't tested it yet.

**Comment by Kreestyan** [comment]
> > I feel bad asking but any ETA for an alpha version? Then again, we can help for testing – as our implementation is blocked for now.
> 
> I don't think this will make to 1.11. But if I made it in a branch of mine could you test it?

I can definitely volunteer to test it. Will I have to build a local dapr or can I install an alpha version with helm? If the former, I may need some directions to build the project and and have it running and tested on my local laptop,
 
> > _which BTW I think this example has a typo since it shows 2 ‘**?**’ where the 2nd one should be replaced by ‘**&**’_
> 
> Could you cut an issue under dapr/docs for this and cc me to it? I think there are multiple issues with that document honestly:

I created a new issue [#3353](https://github.com/dapr/docs/issues/3353)

Thank you!

**Comment by Kreestyan** [comment]
> So it could be something as simple as #2808, which is available on my branch on https://github.com/tmacam/dapr-components-contrib/tree/VaultNamespace. Now: I haven't tested it yet.

I just had a quick look at the changes, they look exactly to what I was expecting. I do believe that it is indeed an easy and low risk change as if one does not use that metadata, I don't see how it could break the other ones.

**Comment by Kreestyan** [comment]
Hi All,

Do we have any update on when this change might be available in an alpha release? Our dapr deployment is on hold because of this.

Thank you

**Comment by tmacam** [comment]
@Kreestyan , could you follow https://github.com/dapr/dapr/blob/master/docs/development/developing-dapr.md and deploy the changes from my branch in your test cluster to verify this works? This vault is a stable component, I can't submit a half-done feature as a PR for it until it is completed, and I can't complete without proper testing.

**Comment by tmacam** [comment]
Also, one additional note to those following this issue, from Vaults documentation on https://developer.hashicorp.com/vault/docs/enterprise/namespaces#usage

> Alternatively, the fully qualified path can be provided without using the X-Vault-Namespace header. In either scenario, Vault will construct the fully qualified path from these two sources to correctly route the request to the appropriate namespace.
>
> For example, these three requests are equivalent:
>
> * Path: ns1/ns2/secret/foo
> * Path: secret/foo, Header: X-Vault-Namespace: ns1/ns2/
> * Path: ns2/secret/foo, Header: X-Vault-Namespace: ns1/

So just prepending your namespace to `enginePath` should unblocking you while a proper fix is released. 

In the meantime I would appreciate if anyone could report if the fix I have proposed in my branch/PR works with either enterprise or HCP. Namespaces is not available in vault OSS version so all I can do is "reproduce what they document" sort of tests.

**Comment by Kreestyan** [comment]
> Also, one additional note to those following this issue, from Vaults documentation on https://developer.hashicorp.com/vault/docs/enterprise/namespaces#usage
> 
> > Alternatively, the fully qualified path can be provided without using the X-Vault-Namespace header. In either scenario, Vault will construct the fully qualified path from these two sources to correctly route the request to the appropriate namespace.
> > For example, these three requests are equivalent:
> > 
> > * Path: ns1/ns2/secret/foo
> > * Path: secret/foo, Header: X-Vault-Namespace: ns1/ns2/
> > * Path: ns2/secret/foo, Header: X-Vault-Namespace: ns1/
> 
> So just prepending your namespace to `enginePath` should unblocking you while a proper fix is released.
> 
> In the meantime I would appreciate if anyone could report if the fix I have proposed in my branch/PR works with either enterprise or HCP. Namespaces is not available in vault OSS version so all I can do is "reproduce what they document" sort of tests.

Hi, I tested this, and YES, it worked. So now I don't know... to me it is good enough. Not sure why our Hashicorp SME did not mention that earlier. So happy, to finally get dapr talking to my Hashicorp vault!!! Thank you so much.
